Set TypeScript to strict mode and fix issues related to server types

jbaez opened the pull request:
Added new npm commands to run lint and check code. We would normally just use `npm run check:all` which does all the checks.

Fixed all the lint and TypeScript issues, removing unused imports and variables, adding types and interfaces, and updating the code where needed.

The commits in this PR have more info on the changes made

jbaez committed:
Fix lint issues and some other TS issues

- set TypeScript in strict mode
- add npm commands to lint / check code
- fix all lint issues
- fix some TS issues
- rename User reponse DTO to make it consistent with the other ones
- override Express/User interface to use UserResponseDto interface
 This is for when the accessing the `user` from a Express Request,
 like in `asset-upload-config`

jbaez committed:
Fix the rest of TS issues

- fix all the remaining TypeScript errors
- add missing `@types/mapbox__mapbox-sdk` package

jbaez committed:
Merge remote-tracking branch 'origin/main' into fix/server-types

alextran1502 commented on server/apps/immich/src/config/asset-upload.config.ts:
```diff
@@ -20,13 +20,18 @@ export const assetUploadOption: MulterOptions = {
 
       if (file.fieldname == 'assetData') {
-        const originalUploadFolder = `${basePath}/${req.user['id']}/original/${req.body['deviceId']}`;
+        const originalUploadFolder = `${basePath}/${req.user.id}/original/${req.body['deviceId']}`;
```
So I got this error when trying to run the app

<img width="879" alt="Screen Shot 2022-06-24 at 22 01 46" src="https://user-images.githubusercontent.com/27055614/175755787-1a99d338-0230-45ed-b91b-06b328b7ea82.png">

And the type of `user` here is `User` from `Express.User`
<img width="862" alt="Screen Shot 2022-06-24 at 22 01 40" src="https://user-images.githubusercontent.com/27055614/175755796-61b75e40-3ab3-463b-8bac-2c2d321ac2d0.png">

alextran1502 replied:
The user here is returned from the function `validate` in `jwt.strategy.ts` which is a UserEntit
<img width="630" alt="Screen Shot 2022-06-24 at 22 18 19" src="https://user-images.githubusercontent.com/27055614/175756182-ce8c9b18-880e-4134-a0a2-a24fd2b1695e.png">
y

alextran1502 replied:
Here is my proposed fix for this issue in `asset-upload.config.ts` and `profile-image-upload.config.ts`

<img width="878" alt="image" src="https://user-images.githubusercontent.com/27055614/175756265-e0582470-89a1-4eb4-a912-7613eee16c49.png">

jbaez replied:
Yes sorry about that, the e2e test was working in my computer (this is also the reason why e2e tests failed in CI) because the container was already created and still  had the old `tsconfig` inside, so no strict mode 😅.

I have added a much neater fix, but would probably need to move the `global.d.ts` somewhere inside src, I'll take a look at fixing this.
Using `as` should be the last resort in TypeScript (also `any` but this is a subject for another day 😄)

I had added a short explanation in the git commit message:
```
override Express/User interface to use UserResponseDto interface
     This is for when the accessing the `user` from a Express Request,
     like in `asset-upload-config`
```

So Express left the `User` interface empty, so you can override it with your own interface like
```
declare global {
  namespace Express {
    // eslint-disable-next-line @typescript-eslint/no-empty-interface
    interface User extends UserResponseDto {}
  }
}
```
I've used `UserResponseDto` for now, although we can create a specific interface for the user returned in `jwt.strategy`

alextran1502 reviewed (changes requested):
Thank you for another high-quality PR to improve the code base! I've left some comments and an idea to fix some issues while running on the dev server. Let me know what do you think

jbaez commented on server/libs/database/src/entities/user.entity.ts:
```diff
@@ -3,32 +3,32 @@ import { Column, CreateDateColumn, Entity, PrimaryGeneratedColumn } from 'typeor
 export class UserEntity {
   @PrimaryGeneratedColumn('uuid')
-  id: string;
+  id!: string;
```
Yes exactly, it's called definite assignment assertion. In general we should only be using these in Entities where properties are not null, because TypeORM cannot use a constructor to initialise the values. So we know that after the Entity is "hydrated" by TypeORM, those properties would have its value.

jbaez commented on server/apps/immich/src/api-v1/album/album-repository.ts:
```diff
@@ -39,7 +39,7 @@ export class AlbumRepository implements IAlbumRepository {
   ) {}
 
   async create(ownerId: string, createAlbumDto: CreateAlbumDto): Promise<AlbumEntity> {
-    return await getConnection().transaction(async (transactionalEntityManager) => {
```
It doesn't, I just removed it because in these case is not useful. When using await what it does is, it resolves the promise, so we now have the "album entity", then it wraps it in another resolved promise which is then returned.

Without the await it just returns the promise. A good practice is to have the return values defined in the function (don't let TypeScript infer), in this case `Promise<AlbumEntity>`, so whatever you do inside you must still return that type or TypeScript will complain.

jbaez commented on server/apps/immich/src/api-v1/album/album.service.spec.ts:
```diff
@@ -25,6 +25,7 @@ describe('Album service', () => {
     albumEntity.createdAt = 'date';
     albumEntity.sharedUsers = [];
     albumEntity.assets = [];
+    albumEntity.albumThumbnailAssetId = null;
```
The entity has defined `albumThumbnailAssetId` as `nullable: true`. So when TypeORM "hydrated" the entity, it would set it to `null` or `string`. This spec was not setting `albumThumbnailAssetId` to anything (`undefined`), so the type was not matching what TypeORM would "hydrate" it to, and TypeScript was complaining because I used `undefined` (instead of null) later in one of the tests for checking the expectation.

jbaez commented on server/apps/immich/src/api-v1/asset/asset.service.ts:
```diff
@@ -166,6 +194,10 @@ export class AssetService {
           'Content-Type': 'image/jpeg',
         });
+        if (!asset.resizePath) {
+          Logger.error('Error serving IMAGE asset for web ');
```
I based those lines in the ones below in line 198-199:

```typescript
Logger.error('Error serving IMAGE asset ', e);
throw new InternalServerErrorException(`Failed to serve image asset ${e}`, 'ServeFile');
```
Looks like all `InternalServerErrorException` in this file use `ServeFile` as the second param. So I think it is consistent with the other ones.

Agree, we could also add `ServeFile` as the second argument of `Logger.error`, since we don't have an error to pass as in there (like in the other places).

Let me know what you think

jbaez commented on server/apps/immich/src/api-v1/asset/asset.service.ts:
```diff
@@ -44,10 +55,14 @@ export class AssetService {
     asset.isFavorite = assetInfo.isFavorite;
     asset.mimeType = mimeType;
-    asset.duration = assetInfo.duration;
+    asset.duration = assetInfo.duration || null;
```
OK, then the best place to fix it would be in the `AssetResponseDto`. So this line would correct, because that's just how it's saved in the DB. And I'll update the `AssetResponseDto` so `duration` is of type `string`, and when we map the entity to the DTO, if `duration` is null, it would return `0:00:00.00000`

jbaez commented on server/apps/immich/src/modules/background-task/background-task.processor.ts:
```diff
@@ -29,11 +28,14 @@ export class BackgroundTaskProcessor {
-          console.log('error deleting ', asset.originalPath);
-        }
-      });
+      // TODO: what if there is no asset.resizePath. Should fail the Job?
```
Before I put the `if`, when `asset.resizePath` was undefined it would probably throw an error. Since I made it safe adding the `if`, it would not longer  throw an error.

I imagine the Job might fail if there is an unhandled error thrown 🤔. An error from the callback of `fs.unlink` doesn't seem like it would make the Job fail either. Maybe it should use something like `job.moveToFailed()`, or maybe just ignoring it it's fine in this case 😄

jbaez commented on server/apps/immich/src/app.controller.ts:
```diff
@@ -2,5 +2,10 @@ import { Controller, Get, Res, Headers } from '@nestjs/common';
 export class AppController {
-  constructor() {}
+  @Get()
+  async redirectToWebpage(@Res({ passthrough: true }) res: Response, @Headers() headers: Record<string, string>) {
```
Oops, sorry by bad. I had a merge conflict which I solved the wrong way.

jbaez commented on server/apps/microservices/src/processors/metadata-extraction.processor.ts:
```diff
@@ -128,7 +127,7 @@ export class MetadataExtractionProcessor {
       }
     } catch (error) {
-      Logger.error(`Failed to trigger object detection pipe line ${error.toString()}`);
+      Logger.error(`Failed to trigger object detection pipe line ${String(error)}`);
```
🙂 it's handy in this case, where error could be anything, not just an `Error` instance with a `toString` method.

jbaez committed:
Move global.d.ts to server `src` folder

jbaez committed:
Update AssetReponseDto duration type

This is now of type `string` that defaults to '0:00:00.00000' if not set
which is what the mobile app currently expects

jbaez committed:
Set context when logging error in asset.service

Use `ServeFile` as the context for logging an error when
asset.resizePath is not set

jbaez committed:
Fix wrong AppController merge conflict resolution

`redirectToWebpage` was removed in main as is no longer used.

jbaez commented:
> Thank you for another high-quality PR to improve the code base! I've left some comments and an idea to fix some issues while running on the dev server. Let me know what do you think

My pleasure. I've fixed the issues and replied the comments. Let me know your thoughts.
